In [5]:
import pandas as pd
import json
import sys
stdout = sys.stdout
reload(sys)
sys.setdefaultencoding('utf-8')
sys.stdout = stdout

----

### Use the following cell to find errors in original data

In [3]:
# read json file
with open('ufo_awesome.json') as data_file:
#     data = json.load(data_file)
    data = []
    errors = 0
    for line in data_file:
        try:
            data.append(json.loads(line))
        except ValueError:
            errors += 1
#             error_line = line
            print line
            continue
data_file.close()

----

In [40]:
df = pd.read_json('ufo_awesome.json',lines=True)

In [41]:
df.shape

(61066, 6)

In [42]:
# drop duplicates
df = df.drop_duplicates()

In [43]:
df.shape

(61050, 6)

You can see that 16 duplicates were removed

----

The shape column is left blank in many instances, so we replace all blank entries with 'unknown'

In [44]:
df['shape'].unique()

array([u'', u' cone', u' light', u' triangle', u' fireball', u' unknown',
       u' rectangle', u' sphere', u' delta', u' disk', u' other', u' dome',
       u' changed', u' round', u' diamond', u' teardrop', u' chevron',
       u' oval', u' cylinder', u' cigar', u' changing', u' egg',
       u' circle', u' hexagon', u' formation', u' crescent', u' pyramid',
       u' flare', u' cross', u' flash'], dtype=object)

In [48]:
df[df['shape'].str.len() == 0].shape

(2512, 6)

There are 2512 rows without any reported shape -- so we assign the string 'unknown' to such rows

In [50]:
df['shape'] = df['shape'].replace('', 'unknown')

In [51]:
df['shape'].unique()

array(['unknown', u' cone', u' light', u' triangle', u' fireball',
       u' unknown', u' rectangle', u' sphere', u' delta', u' disk',
       u' other', u' dome', u' changed', u' round', u' diamond',
       u' teardrop', u' chevron', u' oval', u' cylinder', u' cigar',
       u' changing', u' egg', u' circle', u' hexagon', u' formation',
       u' crescent', u' pyramid', u' flare', u' cross', u' flash'], dtype=object)

#### end of cleaning -- write to csv

In [52]:
df.to_csv('reference.csv',encoding='utf-8')